### PR TITLE
Sdk lightning actions

### PR DESCRIPTION
**Pull request by D4nte**
Proof that https://github.com/comit-network/comit-js-sdk/pull/141 works.

Adds a test that does the flow as expected by HALight using hold invoices.

Ready for review.

### CODE REVIEW

File: api_tests/lib/actors/actor.ts [hunk 1]
```diff
@@ -849,18 +872,84 @@ export class Actor {
...
+        await this.wallets.lightning.pay(request);
+    }
+
+    public async assertLnInvoiceSettled(secretHash: string) {
+        const resp = await this.wallets.lightning.lookupInvoice(secretHash);
+        this.logger.debug(
+            `Checking if invoice is settled, status is: ${resp.state}`
+        );
+        if (resp.state !== 1) {
+            // This is InvoiceState.SETTLED from lnd-async. Hardcoding value until type definition is sorted.
```
Comment (D4nte) [inline api_tests/lib/actors/actor.ts]: Need to fix that.

File: api_tests/lib/actors/actor.ts [hunk 2]
```diff
@@ -849,18 +872,84 @@ export class Actor {
         }
     }
 
-    public async createLnInvoice(sats: number) {
-        return this.wallets.lightning.createInvoice(sats);
+    public lnCreateSha256Secret(): { secret: string; secretHash: string } {
```
Comment (D4nte) [inline api_tests/lib/actors/actor.ts]: Need to add comment to say it's to be removed once cnd supports lightning

File: api_tests/sanity/hold_lnd_payment.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    twoActorTest(
+        "sanity-lnd-alice-pays-bob-using-hold-invoice",
+        async function({ alice, bob }) {
+            await alice.sendRequest(
+                { ledger: LedgerKind.Lightning, asset: AssetKind.Bitcoin },
+                { ledger: LedgerKind.Bitcoin, asset: AssetKind.Bitcoin }
+            );
+
+            const satAmount = "10000";
+            const finalCltvDelta = 40; // This MUST be 40 to match Bob's invoice. Further investigation needed.
```
Comment (D4nte) [inline api_tests/sanity/hold_lnd_payment.ts]: I propose to investigate that in a follow-up PR as we would need to play with `AddHoldInvoice` and understand whether we can manipulate the `finalCtlvDelta` of an invoice when adding it or whether it's a global lnd parameter.

File: api_tests/sanity/hold_lnd_payment.ts [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+            const { secret, secretHash } = bob.lnCreateSha256Secret();
+            await bob.lnCreateHoldInvoice(satAmount, secretHash, 3600);
+            const paymentPromise = alice.lnSendPayment(
+                bob,
+                satAmount,
+                secretHash,
+                finalCltvDelta
+            );
+            await sleep(1000); // Should actually check on bob side if the payment is available
```
Comment (D4nte) [inline api_tests/sanity/hold_lnd_payment.ts]: Need to look if I can fix that.

File: api_tests/lib/actors/actor.ts [hunk 5]
```diff
@@ -849,18 +873,98 @@ export class Actor {
...
-        if (!resp.is_confirmed) {
-            throw new Error(`Invoice ${id} is not confirmed}`);
+    public async lnSendPayment(
+        to: Actor,
+        satAmount: string,
+        secretHash: string,
+        finalCltvDelta: number
+    ) {
+        const toPubkey = await to.wallets.lightning.inner.getPubkey();
+        this.logger.debug(
```
Comment (thomaseizinger) [inline api_tests/lib/actors/actor.ts]: Have you verified that these log statements output what you think they output? I am not so sure here :D

File: api_tests/lib/actors/actor.ts [hunk 6]
```diff
@@ -849,18 +873,98 @@ export class Actor {
...
+        this.logger.debug("LN: Send Payment Response:", resp);
+        return resp;
+    }
+
+    public async lnSettleInvoice(secret: string, secretHash: string) {
+        try {
+            await this.lnAssertInvoiceAccepted(secretHash);
+            this.logger.debug("LN: Settle Invoice", secret, secretHash);
+            await this.wallets.lightning.inner.settleInvoice(secret);
+        } catch {
```
Comment (thomaseizinger) [inline api_tests/lib/actors/actor.ts]: :mega: What is this try-catch logic?

File: api_tests/lib/ledgers/lnd_instance.ts [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+        private readonly logger: Logger,
+        private readonly testLogDir: string,
+        private readonly actorConfig: E2ETestActorConfig,
+        private readonly bitcoindDataDir: string
+    ) {}
+
+    public async start() {
+        const bin = process.env.LND_BIN ? process.env.LND_BIN : "lnd";
+
+        this.logger.debug(`[${this.actorConfig.name}] using binary ${bin}`);
```
Comment (thomaseizinger) [inline api_tests/lib/ledgers/lnd_instance.ts]: The actor name should already be in the logger?

File: api_tests/lib/ledgers/lnd_instance.ts [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+        this.lndDir = path.join(
+            this.testLogDir,
+            "lnd-" + this.actorConfig.name
+        );
+        await mkdirAsync(this.lndDir, "755");
+        await this.createConfigFile(this.lndDir);
```
Comment (thomaseizinger) [inline api_tests/lib/ledgers/lnd_instance.ts]: If you are calling a fn on `this`, you don't need to pass arguments :)

File: api_tests/lib/ledgers/lnd_instance.ts [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+
+        this.logger.debug("Waiting for lnd log file to exist:", this.logPath());
+        await waitUntilFileExists(this.logPath());
+
+        this.logger.debug("Waiting for lnd password RPC server");
+        await this.logReader().waitForLogMessage(
+            "RPCS: password RPC server listening"
+        );
+
+        {
```
Comment (thomaseizinger) [inline api_tests/lib/ledgers/lnd_instance.ts]: Can we have these chunks moved into functions?

File: api_tests/sanity/hold_lnd_payment.ts (hunk 3 above)
Comment (D4nte) [reply api_tests/sanity/hold_lnd_payment.ts]: Ok, this is now resolved locally. The `cltvExpiry` of `addHoldInvoice`  needs to match the `finalCltvDelta` on `sendPayment` when paying the invoice.

Which means that we need to add this `cltvExpiry` to the `ln-add-hold-invoice` action payload.

File: api_tests/lib/ledgers/lnd_instance.ts [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+    public getLightningHost() {
+        return "127.0.0.1";
+    }
+
+    public getLightningPort() {
+        return this.actorConfig.lndP2pPort;
+    }
+
+    private async createConfigFile(lndDir: string) {
+        // We don't use REST but want a random port so we don't get used port errors.
```
Comment (thomaseizinger) [inline api_tests/lib/ledgers/lnd_instance.ts]: You could disable the interface as-well? :)

File: api_tests/lib/wallets/lightning.ts [hunk 11]
```diff
@@ -136,28 +148,26 @@ export class LightningWallet implements Wallet {
...
-            this.logger.debug("Found a channel:", channel);
-            if (
-                channel.transaction_id === transactionId &&
-                channel.transaction_vout === transactionVout
-            ) {
-                return;
+        if (channels) {
+            for (const channel of channels) {
+                this.logger.debug(`Looking for channel ${txId}:${vout}`);
+                this.logger.debug("Found a channel:", channel);
```
Comment (thomaseizinger) [inline api_tests/lib/wallets/lightning.ts]: I think this log statement should be inside the `if`? :D

File: api_tests/sanity/hold_lnd_payment.ts (hunk 3 above)
Comment (thomaseizinger) [reply api_tests/sanity/hold_lnd_payment.ts]: Please open an issue.

File: api_tests/lib/actors/actor.ts (hunk 5 above)
Comment (D4nte) [reply api_tests/lib/actors/actor.ts]: Oh yeah absolutely, it has been key in debugging some hiccups:
`[2020-03-04T10:37:13.228] [DEBUG] alice - LN: Send Payment - to: 03e8be6f8ccc7574402aa750de5fd0737de3aeca5dbaf474798449a72e038e298e ; amt: 10000 ; hash: e35d2ed167ea0b9dae31c88afa270b5714d73920e58e30cc645cb3002848ea4f ; finalCltvDelta:  30`

File: api_tests/lib/actors/actor.ts (hunk 6 above)
Comment (D4nte) [reply api_tests/lib/actors/actor.ts]: Explanatory comment added.

File: api_tests/lib/ledgers/lnd_instance.ts (hunk 10 above)
Comment (tcharding) [reply api_tests/lib/ledgers/lnd_instance.ts]: We just wrote code in cnd to hit lnd via the REST endpoint.  We have a mis-match here, I assume we are going to have e2e tests eventually that hit the REST endpoint also in order to mimic things we do in cnd?

Comment (D4nte) [reply api_tests/lib/ledgers/lnd_instance.ts]: No, we unfortunately cannot ☹️

Comment (thomaseizinger) [reply api_tests/lib/ledgers/lnd_instance.ts]: Fair enough yes, what was I thinking :D

File: api_tests/sanity/hold_lnd_payment.ts (hunk 3 above)
Comment (D4nte) [reply api_tests/sanity/hold_lnd_payment.ts]: Draft PR is coming :)

Comment (D4nte) [reply api_tests/sanity/hold_lnd_payment.ts]: See #2155

**Pull request merged**